### PR TITLE
Introduce Azure Accelerated Networking

### LINKED ISSUE #55: Accelerated Network Enablement
**What would you like to be added**:
Azure supports a feature called Accelerated Networking to increase network performance of virtual machines e.g. reduce latency/more packet per second, lower cpu utilization.
The usage of the feature depends on the machine type and the operating systems.
More information are available [here](https://docs.microsoft.com/en-us/azure/virtual-network/create-vm-accelerated-networking-cli).

**Why is this needed**:
This feature comes for free and should therefore be enabled  by default for all machines and os images combinations which support it.

To enable the feature we just need to turn on a flag on the network interfaces attached to the vm (see [here](https://github.com/Azure/azure-sdk-for-go/blob/master/services/network/mgmt/2019-11-01/network/models.go#17582)).

We would need to put the information which machine type and os image support accelerated networking into the Azure specific part of the CloudProfile. The supported machine types can be fetched via an Azure api (e.g. `az vm list-skus`) and therefore it can be automated.
The operating systems need to be maintained manually. A list is of supported ones is available [here](https://docs.microsoft.com/en-us/azure/virtual-network/create-vm-accelerated-networking-cli#supported-operating-systems).
Custom images are also supported but there are some requirements, see [here](https://docs.microsoft.com/en-us/azure/virtual-network/create-vm-accelerated-networking-cli#custom-images).

**Action Items**
- [x] MCM need to support creating machine with network interfaces that have accelerated networking enabled.
- [ ] Gardener Azure Provider need to render and deploy proper machine classes
- [ ] Gardener Azure CloudProfile need to be extended to allow to configure accelerated networking supporting machine types and operating systems

Automate CloudProfile creation wherever possible by using the Azure apis.

### PR DESCRIPTION
**Pull request by dkistner**
**What this PR does / why we need it**:
Enable automatically Azure Accelerated Networking for all compatible machines configurations.  The machine type and os image version need to be compatible to make use of Azure Accelerated Networking.

Azure Accelerated Networking provide better networking performance, see [here](https://docs.microsoft.com/en-us/azure/virtual-network/create-vm-accelerated-networking-cli#benefits).

A list of compatible machine types can be maintained in the CloudProfile via `.spec.providerConfig.acceleratedNetworingMachineTypes[]`.
The os image versions which are compatible can be maintained in the CloudProfile via `.spec.providerConfig.machineImages[].versions[].acceleratedNetworking`.

ℹ️ This PR require a mcm release which include https://github.com/gardener/machine-controller-manager/pull/438. This is already integrated with https://github.com/gardener/gardener-extension-provider-azure/pull/60.

**Which issue(s) this PR fixes**:
Fixes #55

**Special notes for your reviewer**:
Create a cluster with compatible machine type and os image version. Ssh to machine and exec instructions [here](https://docs.microsoft.com/en-us/azure/virtual-network/create-vm-accelerated-networking-cli#confirm-that-accelerated-networking-is-enabled).

**Release note**:

```improvement user
Azure Accelerated Networking is now automatically enabled if the machine type and operating system + version is compatible.
```

```action operator
Azure Accelerated Networking can now be used for machines which use a compatible machine type and operating system + version. Operators need to maintain a list of compatible machine types and label compatible os image version in the Azure CloudProfile.
```

### CODE REVIEW

File: docs/usage-as-operator.md [hunk 1]
```diff
@@ -22,17 +19,29 @@ countUpdateDomains:
 countFaultDomains:
 - region: westeurope
   count: 3
+acceleratedNetworkingMachineTypes:
+- Standard_D3_v2
```
Comment (rfranzke) [inline docs/usage-as-operator.md:23]:
```suggestion
- name: Standard_D3_v2
```

File: docs/usage-as-operator.md [hunk 2]
```diff
@@ -75,6 +88,8 @@ spec:
   providerConfig:
     apiVersion: azure.provider.extensions.gardener.cloud/v1alpha1
     kind: CloudProfileConfig
+    acceleratedNetworkingMachineTypes:
+    - Standard_D3_v2
```
Comment (rfranzke) [inline docs/usage-as-operator.md:92]:
```suggestion
    - name: Standard_D3_v2
```

File: pkg/apis/azure/types_cloudprofile.go [hunk 3]
```diff
@@ -31,6 +31,8 @@ type CloudProfileConfig struct {
 	// MachineImages is the list of machine images that are understood by the controller. It maps
 	// logical names and versions to provider-specific identifiers.
 	MachineImages []MachineImages
+	// AcceleratedNetworkingMachineTypes is a list of machine types which support accelerated networking.
+	AcceleratedNetworkingMachineTypes []string
 }
```
Comment (rfranzke) [inline pkg/apis/azure/types_cloudprofile.go:36]: To allow further enhancements it's better to use here: 

```suggestion
	// AcceleratedNetworkingMachineTypes is a list of machine types which support accelerated networking.
	AcceleratedNetworkingMachineTypes []AcceleratedNetworkingMachineType
}

// doc string
type AcceleratedNetworkingMachineType struct {
  Name string
}
```

File: pkg/apis/azure/v1alpha1/types_cloudprofile.go [hunk 4]
```diff
@@ -32,6 +32,8 @@ type CloudProfileConfig struct {
 	// MachineImages is the list of machine images that are understood by the controller. It maps
 	// logical names and versions to provider-specific identifiers.
 	MachineImages []MachineImages `json:"machineImages"`
+	// AcceleratedNetworkingMachineTypes is a list of machine types which support accelerated networking.
```
Comment (rfranzke) [inline pkg/apis/azure/v1alpha1/types_cloudprofile.go:35]:
```suggestion
	// AcceleratedNetworkingMachineTypes is a list of machine types which support accelerated networking.
	// +optional
```

File: pkg/apis/azure/v1alpha1/types_cloudprofile.go [hunk 5]
```diff
@@ -60,4 +62,6 @@ type MachineImageVersion struct {
 	// ID is the Shared Image Gallery image id.
 	// +optional
 	ID *string `json:"id,omitempty"`
+	// AcceleratedNetworking is an indicator if the image does support Azure accelerated networking.
```
Comment (rfranzke) [inline pkg/apis/azure/v1alpha1/types_cloudprofile.go:65]:
```suggestion
	// AcceleratedNetworking is an indicator if the image does support Azure accelerated networking.
	// +optional
```

File: pkg/apis/azure/types_cloudprofile.go (hunk 3 above)
Comment (rfranzke) [reply pkg/apis/azure/types_cloudprofile.go:36]: Also, can you implement a validation that only machine types that are also configured in `.spec.machineTypes` are present in this list? Does something like this make sense?

File: pkg/controller/worker/machines.go [hunk 6]
```diff
@@ -268,3 +273,32 @@ func (w *workerDelegate) generateMachineConfig(ctx context.Context) error {
 
 	return nil
 }
+
+func (w *workerDelegate) enableAcceleratedNetworking(machineType, machineImageName, machineImageVersion string) bool {
```
Comment (rfranzke) [inline pkg/controller/worker/machines.go:277]: Can you add a unit test case that uses this accelerated networking?

File: pkg/controller/worker/machines.go [hunk 7]
```diff
@@ -198,12 +199,16 @@ func (w *workerDelegate) generateMachineConfig(ctx context.Context) error {
 				machineClassSpec["vnetResourceGroup"] = *infrastructureStatus.Networks.VNet.ResourceGroup
 			}
 
+			if w.enableAcceleratedNetworking(pool.MachineType, pool.MachineImage.Name, pool.MachineImage.Version) {
```
Comment (rfranzke) [inline pkg/controller/worker/machines.go:202]: WDYT about

```suggestion
			if w.enableAcceleratedNetworking(pool.MachineType, pool.MachineImage) {
```

File: pkg/controller/worker/machines.go [hunk 8]
```diff
@@ -268,3 +273,32 @@ func (w *workerDelegate) generateMachineConfig(ctx context.Context) error {
 
 	return nil
 }
+
+func (w *workerDelegate) enableAcceleratedNetworking(machineType, machineImageName, machineImageVersion string) bool {
+	// Check if used operating system and version are supporting accelerated networking.
+	var osSupport bool
+	for _, image := range w.cloudProfileConfig.MachineImages {
+		if image.Name != machineImageName {
```
Comment (rfranzke) [inline pkg/controller/worker/machines.go:281]: Hm, what if a machine image that is currently in use for shoots is removed from the `CloudProfile`? Then, in the next reconciliation the accelerated networking won't work anymore, right?

File: pkg/controller/worker/machines.go [hunk 9]
```diff
@@ -268,3 +273,32 @@ func (w *workerDelegate) generateMachineConfig(ctx context.Context) error {
...
+			continue
+		}
+		for _, imageVersion := range image.Versions {
+			if imageVersion.Version == machineImageVersion {
+				if imageVersion.AcceleratedNetworking != nil && *imageVersion.AcceleratedNetworking {
+					osSupport = true
+				} else {
+					return false
+				}
```
Comment (rfranzke) [inline pkg/controller/worker/machines.go:290]:
```suggestion
				if imageVersion.AcceleratedNetworking == nil {
					return false
			  }
				osSupport = *imageVersion.AcceleratedNetworking
```

File: docs/usage-as-operator.md (hunk 1 above)
Comment (dkistner) [reply docs/usage-as-operator.md:23]: Hmm yes, we could do that, but it is simply a list of machine type names. We could rename the list key to `acceleratedNetworkingMachineTypeNames` to make it more explicit.

File: pkg/apis/azure/types_cloudprofile.go (hunk 3 above)
Comment (dkistner) [reply pkg/apis/azure/types_cloudprofile.go:36]: Hmm yes, we could do that, but this is basically a list of machine types which support accelerated networking and not the list of usable machine types. These information could be fetched automatically for all machine types e.g. via `az vm list-skus`. The list of usable machine types is most probably maintained manually and only selected machine types are part of it.

I'm also ok with having that validation.

File: pkg/controller/worker/machines.go (hunk 6 above)
Comment (dkistner) [reply pkg/controller/worker/machines.go:277]: Yep. In general I wanted to add more test for the worker creation.

File: pkg/controller/worker/machines.go (hunk 8 above)
Comment (dkistner) [reply pkg/controller/worker/machines.go:281]: Yep, exactly, but existing machine with accelerated networking wouldn't be affected.

File: docs/usage-as-operator.md (hunk 1 above)
Comment (rfranzke) [reply docs/usage-as-operator.md:23]: Today it's a list of machine types. Tomorrow it's maybe more (e.g., additional configuration/information per machine type). Just guessing, of course, though, generally it's API-wise a good practice to plan for such future enhancements. If you also know that something like this is highly unlikely to happen then keep `[]string`, though.

File: pkg/apis/azure/types_cloudprofile.go (hunk 3 above)
Comment (rfranzke) [reply pkg/apis/azure/types_cloudprofile.go:36]: See https://github.com/gardener/gardener-extension-provider-azure/pull/65#discussion_r408030329 - your decision :)

File: pkg/controller/worker/machines.go (hunk 8 above)
Comment (rfranzke) [reply pkg/controller/worker/machines.go:281]: Hm, ok, wouldn't it? What does MCM when the respective field in the `AzureMachineClass` changes? Probably nothing... ok. Fine, resolving the conversation then.

File: docs/usage-as-operator.md (hunk 1 above)
Comment (dkistner) [reply docs/usage-as-operator.md:23]: Hmm yes I agree with you. Then I would suppose to make it even more generic similar we have it for the cloud provider specific information for images.
Let's have a list in the provider section like that.
```
providerConfig:
  machineTypes:
  - name: Standard_XXX
    acceleratedNetworking: true
    premiumStorage: true # Some real property which would make sense for validation purposes.
    someOtherProperty: xxx
  - name: Standard_YYY
    acceleratedNetworking: false
  - name: Standard_ZZZ
```

Comment (rfranzke) [reply docs/usage-as-operator.md:23]: Sure, sounds great 👍

**Comment by dkistner** [comment]
Updated the PR. The requested test is missing. Need some more time to restructure the tests..

File: pkg/apis/azure/types_cloudprofile.go [hunk 10]
```diff
@@ -57,4 +59,14 @@ type MachineImageVersion struct {
...
 	ID *string
+	// AcceleratedNetworking is an indicator if the image does support Azure accelerated networking.
+	AcceleratedNetworking *bool
+}
+
+// MachineType contains provider specific information to a machine type.
+type MachineType struct {
+	// Name is the name of the machine type.
+	Name string
+	// AcceleratedNetworking is an inidicator if the machine type supports Azure accelerated networking.
```
Comment (rfranzke) [inline pkg/apis/azure/types_cloudprofile.go:70]:
```suggestion
	// AcceleratedNetworking is an indicator if the machine type supports Azure accelerated networking.
```

File: pkg/apis/azure/v1alpha1/types_cloudprofile.go [hunk 11]
```diff
@@ -32,6 +32,8 @@ type CloudProfileConfig struct {
 	// MachineImages is the list of machine images that are understood by the controller. It maps
 	// logical names and versions to provider-specific identifiers.
 	MachineImages []MachineImages `json:"machineImages"`
+	// MachineTypes is a list of machine types complete with provider specific information.
```
Comment (rfranzke) [inline pkg/apis/azure/v1alpha1/types_cloudprofile.go:35]:
```suggestion
	// MachineTypes is a list of machine types complete with provider specific information.
	// +optional
```

File: pkg/apis/azure/v1alpha1/types_cloudprofile.go [hunk 12]
```diff
@@ -60,4 +62,15 @@ type MachineImageVersion struct {
 	// ID is the Shared Image Gallery image id.
 	// +optional
 	ID *string `json:"id,omitempty"`
+	// AcceleratedNetworking is an indicator if the image does support Azure accelerated networking.
```
Comment (rfranzke) [inline pkg/apis/azure/v1alpha1/types_cloudprofile.go:65]:
```suggestion
	// AcceleratedNetworking is an indicator if the image does support Azure accelerated networking.
	// +optional
```

File: pkg/controller/worker/machines.go [hunk 13]
```diff
@@ -268,3 +274,35 @@ func (w *workerDelegate) generateMachineConfig(ctx context.Context) error {
...
 }
+
+func (w *workerDelegate) enableAcceleratedNetworking(machineTypeName string, machineImageConfig *extensionsv1alpha1.MachineImage) bool {
+	if machineImageConfig == nil {
+		return false
+	}
+
+	// Check if used operating system and version are supporting accelerated networking.
+	var osSupport bool
+	for _, image := range w.cloudProfileConfig.MachineImages {
```
Comment (rfranzke) [inline pkg/controller/worker/machines.go:285]: There is already a loop over the `MachineImages` here: https://github.com/gardener/gardener-extension-provider-azure/pull/65/files#diff-30e669b862b2c1062573be5e967954c3R128

What about enhancing this function to not only return `urn` and `id` but also this information about the `acceleratedNetworking`? You could pass this as a parameter to this function and then you only need to loop over the machine types below.

Comment (dkistner) [reply pkg/controller/worker/machines.go]: Sounds reasonable for me also as we could also define accelerated networking capabilities in the `machineImage` section of the worker status.

**Comment by dkistner** [comment]
Thanks @rfranzke for the review. 
I have addressed the changes and extended the tests to cover accelerated networking configuration for the workers .

File: pkg/apis/azure/helper/helper.go [hunk 14]
```diff
@@ -102,10 +102,11 @@ func FindImageFromCloudProfile(cloudProfileConfig *api.CloudProfileConfig, image
...
 					return &api.MachineImage{
-						Name:    imageName,
-						Version: version.Version,
-						URN:     version.URN,
-						ID:      version.ID,
+						Name:                  imageName,
+						Version:               version.Version,
+						URN:                   version.URN,
+						ID:                    version.ID,
+						AcceleratedNetworking: version.AcceleratedNetworking,
```
[the remark this answers is not in the document]
Comment (dkistner) [reply pkg/apis/azure/helper/helper.go:109]: done.

**Comment by rfranzke** [comment]
@dkistner I want to merge, but it tells 
>This branch has conflicts that must be resolved

without showing them. Can you check? Maybe it's related to the GitHub issues yesterday evening, can you force-push?

**Comment by rfranzke** [comment]
OK, it seems that this PR was already merged and pushed to `master` as of `357a30d0377385e0bd6f65bd57ee8c0287101d26`, but GitHub didn't mark the PR as merged. Unfortunate.. 👎 I'll then close the PR manually.

**Pull request closed without merging**